### PR TITLE
[Build] Integrate OpenAssetIO

### LINKED ISSUE #15: Integrate OpenAssetIO
Integrate OpenAssetIO with the resolver plugin. This likely means having to build OpenAssetIO manually, and integrating that with CI. Initialise the python plug-in system in the constructor. Configure OpenAssetIO with a manager (BAL almost certainly)
For the moment, a requirement for this resolver is that your enclosing application be being run from a python environment, as there is no C++ manager to facilitate otherwise.

### PR DESCRIPTION
**Pull request by elliotcmorris**
Closes #15 

Add OpenAssetIO as a dependency of the resolver, so that OpenAssetIO methods can be use to implement resolver interface methods.

Add OpenAssetIO building to the test automation along with this. We use the OpenAssetIIO docker container to make getting dependencies for the build easier, and then move the build made inside that container between jobs, (helpfully, our container and the vfxall container are the same base, intentionally, so this is compatible.)

Doing this also means we end up writing an openassetio logger interface implementation, which while it's tested insofar as we know it logs things, it could be tested more thoroughly. However, OpenAssetIO currently only sets severity upon construction, which is problematic to test when it's puppeted by something else.

### CODE REVIEW

**Comment by foundrytom** [comment]
> However, OpenAssetIO currently only sets severity upon construction, which is problematic to test when it's puppeted by something else.

You should be able to set the severity any time you like https://github.com/OpenAssetIO/OpenAssetIO/blob/fbc24bbeb32f6aa0c43cc985310364fa3816bc30/src/openassetio-core/include/openassetio/log/SeverityFilter.hpp#L53

File: src/resolver.cpp [hunk 1]
```diff
@@ -21,103 +27,176 @@ TF_DEBUG_CODES(OPENASSETIO_RESOLVER)
...
+      openassetio::python::hostApi::createPythonPluginSystemManagerImplementationFactory(logger_);
+
+  const auto hostInterface = std::make_shared<UsdOpenAssetIOHostInterface>();
+
+  manager_ = openassetio::hostApi::ManagerFactory::defaultManagerForInterface(
+      hostInterface, managerImplementationFactory, logger_);
+
+  if (!manager_) {
+    throw std::invalid_argument{
+        "OpenAssetIO manger plugin not found. Check OPENASSETIO_DEFAULT_CONFIG is set."};
```
Comment (foundrytom) [inline src/resolver.cpp]: [fix] Can we re-word this slightly (and use the static) as it could be interpreted that a configured plugin was missing, which would actually surface as an execption in `defaultManagerForInterface` iirc. We usually say:

https://github.com/OpenAssetIO/OpenAssetIO/blob/fbc24bbeb32f6aa0c43cc985310364fa3816bc30/examples/host/simpleResolver/simpleResolver.py#L142

**Comment by feltech** [comment]
> > However, OpenAssetIO currently only sets severity upon construction, which is problematic to test when it's puppeted by something else.
> 
> You should be able to set the severity any time you like https://github.com/OpenAssetIO/OpenAssetIO/blob/fbc24bbeb32f6aa0c43cc985310364fa3816bc30/src/openassetio-core/include/openassetio/log/SeverityFilter.hpp#L53

This was more about using the environment variable to be able to control the severity from the (Python) USD application, which has no access to the OpenAssetIO `SeverityFilter` instance.  

The severity is initialised by the env var only on construction of the `SeverityFilter`, after which the env var isn't read again. We did wonder, in passing, if that was a wise design, given e.g. a debug workflow of enabling debug severity just around a specific call, to avoid noise polluting the log output.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -11,24 +11,45 @@ concurrency:
...
       - run: |
           python -m pip install -r tests/requirements.txt
 
       - name: Test
-        run: >
+        run: |
+          export PXR_PLUGINPATH_NAME=$GITHUB_WORKSPACE/build/dist/resources/plugInfo.json
+          export LD_LIBRARY_PATH=/usr/local/lib:$GITHUB_WORKSPACE/openassetio-build/lib64
+          export OPENASSETIO_DEFAULT_CONFIG=$GITHUB_WORKSPACE/tests/resources/openassetio.conf
+          export PYTHONPATH=/usr/local/lib/python/:$GITHUB_WORKSPACE/openassetio-build/lib/python3.9/site-packages
```
Comment (foundrytom) [inline .github/workflows/test.yml:62]: [question] Whats the motivation for adding the system paths in front of the rest for  these? Are they not set by default? If we're just extending, then I think more traditionally it's formulated as:

`export PYTHONPATH=$GITHUB_WORKSPACE/openassetio-build/lib/python3.9/site-packages:$PYTHONPATH`

Which avoids us inadvertently stomping over defaults. If that was the intention, then it'd be good to document the problem in a comment.

File: .github/workflows/code-quality.yml [hunk 3]
```diff
@@ -48,15 +48,20 @@ jobs:
...
+      - name: Build
+        uses: ./.github/build_openassetio
 
   cpp-linters:
     name: C++ linters
     runs-on: ubuntu-20.04
+    needs: build-openassetio
     container:
       image: aswf/ci-vfxall:2022-clang13.1
     strategy:
```
Comment (feltech) [inline .github/workflows/code-quality.yml:63]: [fix]/[plan] `strategy` matrix doesn't make much sense since we're using a container.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -11,24 +11,45 @@ concurrency:
...
           cmake --install build
 
       - run: |
           python -m pip install -r tests/requirements.txt
 
       - name: Test
-        run: >
+        run: |
+          export PXR_PLUGINPATH_NAME=$GITHUB_WORKSPACE/build/dist/resources/plugInfo.json
+          export LD_LIBRARY_PATH=/usr/local/lib:$GITHUB_WORKSPACE/openassetio-build/lib64
```
Comment (feltech) [inline .github/workflows/test.yml:60]: [consider]/[plan] I suspect the fact we're using `root` to build is the reason `/usr/local/lib` isn't on the library search path. We could try to build/run using the `aswfuser` user that is available in the container.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -11,24 +11,45 @@ concurrency:
...
+          export OPENASSETIO_DEFAULT_CONFIG=$GITHUB_WORKSPACE/tests/resources/openassetio.conf
```
Comment (foundrytom) [inline .github/workflows/test.yml:61]: [fix] or [plan] We could do with some test coverage on the handling or manager init as its quite a mission-critical component. ie, dont set this here, and have fixtures that create a known configured environment, such that we can assert the behaviour with there is no default manager set, and when there is.

File: src/resolver.cpp [hunk 6]
```diff
@@ -21,103 +27,176 @@ TF_DEBUG_CODES(OPENASSETIO_RESOLVER)
...
 /* Ar Resolver Implementation */
 UsdOpenAssetIOResolver::UsdOpenAssetIOResolver() {
-  TF_DEBUG(OPENASSETIO_RESOLVER).Msg("OPENASSETIO_RESOLVER: " + TF_FUNC_NAME() + "\n");
+  logger_ =
+      openassetio::log::SeverityFilter::make(std::make_shared<UsdOpenAssetIOResolverLogger>());
+
+  auto managerImplementationFactory =
+      openassetio::python::hostApi::createPythonPluginSystemManagerImplementationFactory(logger_);
+
+  const auto hostInterface = std::make_shared<UsdOpenAssetIOHostInterface>();
```
Comment (foundrytom) [inline src/resolver.cpp:80]: [question] We have the whole `make` thing so that we know we're always using the correct handle types for these things, so the lifetime is properly managed. How do we want to deal with this for dervived classes. Can/should we provide (or require) that sub-classes also use the private constructor/`make` approach too?

File: cmake/DefaultTargetProperties.cmake [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        CXX_STANDARD_REQUIRED ON
+        # Disable compiler extensions. E.g. use -std=c++11  instead of
+        # -std=gnu++11.  Helps limit cross-platform issues.
+        CXX_EXTENSIONS OFF
+    )
+
+    #-------------------------------------------------------------------
+    # Compiler warnings
+
+    include(CompilerWarnings)
```
Comment (feltech) [inline cmake/DefaultTargetProperties.cmake]: [fix]/[plan] For best practices, we should move this `include` outside of this function to the top of the file.

I realise this is also an issue in OpenAssetIO, where this came from, my bad.

We should also remove the duplicate call to `set_default_compiler_warnings` in `src/CMakeLists.txt` as well as the duplicate `include(CompilerWarnings)` from the top-level `CMakeLists.txt`.

Finally, (also an issue in the OpenAssetIO repo) we should namespace `set_default_compiler_warnings` to `openassetio_usdresolver_set_default_compiler_warnings`.

**Review by foundrytom — CHANGES_REQUESTED** [review]
Nice work wrangling all that CI fun! Its amazing now much extra effort there is when we don't have the conan package ready.

Only a couple of fairly minor comments (I'm leaving the CMake stuff to you folks).

As an aside, in more general terms, It's worth thinking of the commit messages as mini DRs too - and focus more on the "why" for particular things (eg: the approach to building OpenAssetIO vs our usual one), vs just a summary of changes, as those are the things we forget later, or raise questions in the review.

File: .github/workflows/test.yml (hunk 2 above)
Comment (elliotcmorris) [reply .github/workflows/test.yml:62]: It's because the vfxall platform has a somewhat odd installation of python/USD.

In the `/usr/local/lib` directory, you'll find a `python3.9` dir, which has all the regular python libs in it, which is on the pythonpath. However, adjacent to that is this `python` directory, which contains only the `pxr` library, which isn't on the pythonpath by default.

File: CMakeLists.txt [hunk 8]
```diff
@@ -20,12 +24,43 @@ endif ()
 
 add_compile_options(-Wno-deprecated)  # This is just to avoid some spammy warnings
```
Comment (feltech) [inline CMakeLists.txt]: [fix]/[plan] Can remove this now that it is set per-target.

File: src/resolver.cpp [hunk 9]
```diff
@@ -21,103 +27,176 @@ TF_DEBUG_CODES(OPENASSETIO_RESOLVER)
...
 PXR_NAMESPACE_CLOSE_SCOPE
 
+namespace {
+/// Converter logger from OpenAssetIO log framing to USD log outputs.
+class UsdOpenAssetIOResolverLogger : public openassetio::log::LoggerInterface {
+ public:
+  void log(Severity severity, const openassetio::Str &message) override {
+    switch (severity) {
+      case Severity::kCritical:
+        TF_ERROR(TfDiagnosticType::TF_DIAGNOSTIC_FATAL_ERROR_TYPE, message + "\n");
```
Comment (feltech) [inline src/resolver.cpp]: [fix]/[plan] Pretty sure we discovered this was adding an extra newline, i.e. the `\n` isn't necessary here (and for some of the other severities).

**Comment by foundrytom** [comment]
> The severity is initialised by the env var only on construction of the `SeverityFilter`, after which the env var isn't read again. We did wonder, in passing, if that was a wise design, given e.g. a debug workflow of enabling debug severity just around a specific call, to avoid noise polluting the log output.

Do you mean in general, or in this instance? More generally, it is the host that decides how logging should be filtered vs OpenAssetIO  - `SeverityFilter` is just a convenience class that provides a default mechanism for simple hosts that dont want to have to think about it too much. I dont think we have an expectation that it is always used. If a host supports run-time logging severity changes, then we'd encourage them to implement a logger interface that is coupled to those in a way that makes sense to the user.

File: .github/workflows/test.yml (hunk 2 above)
Comment (foundrytom) [reply .github/workflows/test.yml:62]: Right, got you, can we add that as a comment then so we know why (and is it supposed to be in front of ours?)

Comment (foundrytom) [reply .github/workflows/test.yml:62]: Same for `LD_LIBRARY_PATH`?

File: src/resolver.h [hunk 10]
```diff
@@ -44,4 +49,8 @@ class UsdOpenAssetIOResolver final : public PXR_NS::ArDefaultResolver {
       const PXR_NS::ArResolvedPath &resolvedPath, WriteMode writeMode) const final;
 
  private:
+  /// @brief OpenAssetIO Logger. Uses TF Log/Error functions.
```
Comment (feltech) [inline src/resolver.h]: [consider] I think the `@brief` is redundant here (and below)?

File: tests/test_resolver.py [hunk 11]
```diff
@@ -21,7 +21,7 @@
 # This test can be removed once the logging transforms, alchemy like,
 # into real functionality.
 def test_open_stage_and_logging(capfd):
-    Usd.Stage.Open("resources/empty_shot.usda")
+    open_stage("resources/empty_shot.usda")
```
Comment (feltech) [inline tests/test_resolver.py:24]: [consider] As discussed offline, would be good to pull this out as a separate commit, since it addresses a bug in a previous PR.

**Review by feltech — COMMENTED** [review]
Most of my comments I've added `/[plan]` to since they can be pulled out as separate tidyup issues if desired.

File: tests/test_resolver.py (hunk 11 above)
Comment (elliotcmorris) [reply tests/test_resolver.py:24]: I think this is in a separate commit already, did you mean a seperate pull request?

File: .github/workflows/test.yml (hunk 2 above)
Comment (elliotcmorris) [reply .github/workflows/test.yml:62]: `LD_LIBRARY_PATH` is different but similar. David found this post
https://unix.stackexchange.com/questions/115129/why-does-root-not-have-usr-local-in-path

Potentially we could do `sudo -u aswfuser python -m pytest`, but i'm not sure I see that as altogheter much of a different approach, and the fact that it's finally working now makes me hesitant to jump back into path wrangling.

Will for sure add comments explaining these two things.

**Comment by feltech** [comment]
> > The severity is initialised by the env var only on construction of the `SeverityFilter`, after which the env var isn't read again. We did wonder, in passing, if that was a wise design, given e.g. a debug workflow of enabling debug severity just around a specific call, to avoid noise polluting the log output.
> 
> Do you mean in general, or in this instance? More generally, it is the host that decides how logging should be filtered vs OpenAssetIO - `SeverityFilter` is just a convenience class that provides a default mechanism for simple hosts that dont want to have to think about it too much. I dont think we have an expectation that it is always used. If a host supports run-time logging severity changes, then we'd encourage them to implement a logger interface that is coupled to those in a way that makes sense to the user.

That's a good point.  It is really just an issue in this instance (and I guess similar situations), where the outer application has no direct visibility or control over the OpenAssetIO host functions.  Probably a rare case, and if we become really bothered by it, like you say, we could write something similar to `SeverityFilter` but customised just for this project's use cases.

File: .github/workflows/test.yml (hunk 2 above)
Comment (elliotcmorris) [reply .github/workflows/test.yml:62]: > and is it supposed to be in front of ours?

I didn't think the ordering would matter, but if it's traditional to put the more application focused ones in the front of the path i'll just swap em around.

Comment (foundrytom) [reply .github/workflows/test.yml:62]: Things in front are loaded in preference to ones on the back, when the same thing is found, so it depends on what the required behaviour there is, usually its that the stuff you're adding is more important.

File: tests/test_resolver.py (hunk 11 above)
Comment (foundrytom) [reply tests/test_resolver.py:24]: Im happy that it's in a separate commit in this PR (the granularity of the other changes would be more worthy of effort than PR juggling here imho), but we shouldn't hold up progress on the functional aspects, as long as we have `plan` for the main things we want to improve for an MVP

File: .github/workflows/test.yml (hunk 4 above)
Comment (elliotcmorris) [reply .github/workflows/test.yml:60]: All things being equal, i'd rather leave this as it stands, rather than opening up this specific can of worms again. I'll write a comment explaining why this is here.

File: .github/workflows/test.yml (hunk 5 above)
Comment (elliotcmorris) [reply .github/workflows/test.yml:61]: Have made an issue for this, deferring to [plan]
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/issues/17

File: src/resolver.cpp (hunk 6 above)
Comment (elliotcmorris) [reply src/resolver.cpp:80]: That makes sense to me personally. Construction consistency could easily be a big point of confusion, especially if it's important that these objects are stored inside `shared_ptrs`. Folks (like myself,) might default to trying to construct objects on the stack if they havn't read the doc.

File: src/resolver.h (hunk 10 above)
Comment (elliotcmorris) [reply src/resolver.h]: Aye, habit, sorry.
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/4cf93450fdc3009570ede596c1a93bc78decdf92

File: src/resolver.cpp (hunk 6 above)
Comment (foundrytom) [reply src/resolver.cpp:80]: Cool, Ill add something in to the OpenAssetIO work so we can come up with a proper plan.

File: src/resolver.cpp (hunk 9 above)
Comment (elliotcmorris) [reply src/resolver.cpp]: Yeah, did my own test and confirmed TF_ERROR and TF_WARN add newlines implicitly.

https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/25bb4b99a864b2be5e7e46798cc14ad76d1a92d7

File: CMakeLists.txt (hunk 8 above)
Comment (elliotcmorris) [reply CMakeLists.txt]: https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/135c8b8aa252f19af13afced0e83fc1088db37fc

(part of another fixup)

File: cmake/DefaultTargetProperties.cmake (hunk 7 above)
Comment (elliotcmorris) [reply cmake/DefaultTargetProperties.cmake]: https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/135c8b8aa252f19af13afced0e83fc1088db37fc

File: .github/workflows/code-quality.yml (hunk 3 above)
Comment (elliotcmorris) [reply .github/workflows/code-quality.yml:63]: https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/55bc8822f4c95b01b973a0905e768468531d30b8

File: src/resolver.cpp (hunk 6 above)
Comment (foundrytom) [reply src/resolver.cpp:80]: https://github.com/OpenAssetIO/OpenAssetIO/issues/808

File: cmake/DefaultTargetProperties.cmake (hunk 7 above)
Comment (elliotcmorris) [reply cmake/DefaultTargetProperties.cmake]: namespace : https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/c8343031d8508ebb5d528a2d63b8848eab9830f9

File: tests/test_resolver.py (hunk 11 above)
Comment (feltech) [reply tests/test_resolver.py:24]: Sorry my bad, it is indeed already in a separate commit :facepalm:

File: cmake/DefaultTargetProperties.cmake (hunk 7 above)
Comment (feltech) [reply cmake/DefaultTargetProperties.cmake]: Are we leaving the name `set_default_compiler_warnings` for now (i.e. not namespacing with `openassetio_usdresolver_`)?  Happy to punt that if it helps get this over the line (its not a new issue, in that it was copied from the OpenAssetIO repo like that).

Comment (foundrytom) [reply cmake/DefaultTargetProperties.cmake]: Can we track it and have a clean-up pass, once we have base line functionality sorted.

File: src/resolver.cpp (hunk 1 above)
Comment (elliotcmorris) [reply src/resolver.cpp]: Took the same pattern, this what you were thinking?

https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/87ca72d3a5937f0bf3245513dd0fe536eb05f3b1

File: .github/workflows/test.yml (hunk 2 above)
Comment (elliotcmorris) [reply .github/workflows/test.yml:62]: Add explainer comment: https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/16/commits/470161404485a2a1a4a0b190e4a56f1b13256d10

File: cmake/DefaultTargetProperties.cmake (hunk 7 above)
Comment (elliotcmorris) [reply cmake/DefaultTargetProperties.cmake]: I did renamespace it in https://github.com/OpenAssetIO/usdOpenAssetIOResolver/commit/c8343031d8508ebb5d528a2d63b8848eab9830f9

Perhaps you'd loaded the review page before that comment appeared, since i forgot to do it and added a subsequent fixup.

**Review by foundrytom — APPROVED** [review]
LGTM, thanks! Be good to add a little context on the reasons for building OpenAssetIO as we do into the commit message when you do the squash.
